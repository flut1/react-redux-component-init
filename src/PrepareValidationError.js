import ExtendableError from './utils/ExtendableError';

class PrepareValidationError extends ExtendableError {}

export default PrepareValidationError;
