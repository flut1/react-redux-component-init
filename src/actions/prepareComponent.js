import { MODE_PREPARE } from '../initMode';
import extractValuesForProps from '../utils/extractValuesForProps';

import initComponent from './initComponent';

/**
 * Action that prepares a single component. Duplicate calls with exactly the same
 * `Component` and `props` will be ignored.
 *
 * @param {react.Component} Component The component to prepare. If the component was not
 * wrapped in `withInitAction`, this action will have no effect and returns a Promise that
 * resolves immediately.
 * @param {object} [props={}] The values of `initProps` to initialize this component with. These
 * values should be the same as the props that the component will eventually mount with. Omitting
 * one of the `initProps` values configured in `withInitAction()` will result in an error.
 * @returns {function} A thunk function that should be passed directly to the Redux `dispatch`
 * function.
 * @example dispatch(prepareComponent(Comment, { commentId: 5 }));
 */
export default (Component, props = {}) =>
  (dispatch, getState) => {
    // unwrap other HoC if no initConfig is found
    while (!Component.initConfig && Component.WrappedComponent) {
      Component = Component.WrappedComponent; // eslint-disable-line no-param-reassign
    }

    if (Component.initConfig) {
      const {
        componentId,
        initProps,
        options: { getInitState, getPrepareKey },
      } = Component.initConfig;

      const initState = getInitState(getState());
      if (!initState) {
        throw new ReferenceError('Could not find init state. Did you attach the init reducer?');
      }

      const { mode } = initState;
      if (mode === MODE_PREPARE) {
        const initValues = extractValuesForProps(props, initProps);

        initValues.forEach((initValue, index) => {
          if (typeof initValue === 'undefined') {
            throw new ReferenceError(`Component "${componentId}" expected prop "${initProps[index]}" but was not passed to prepareComponent`);
          }
        });

        const prepareKey = getPrepareKey(componentId, initValues);

        return dispatch(initComponent(Component, initValues, prepareKey, {
          isPrepare: true,
        }));
      }
    }

    return Promise.resolve();
  };
