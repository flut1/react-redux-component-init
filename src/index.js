export { default as withInitAction } from './withInitAction';
export { default as initReducer } from './reducer';

export { default as initComponent } from './actions/initComponent';
export { default as prepareComponent } from './actions/prepareComponent';
export { default as prepareComponents } from './actions/prepareComponents';
export { default as setInitMode } from './actions/setInitMode';

export { default as PrepareValidationError } from './PrepareValidationError';
export * from './initMode';
export * from './initSelfMode';
export * as actionTypes from './actions/actionTypes';
