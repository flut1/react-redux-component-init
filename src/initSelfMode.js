/**
 * Possible values for the `selfInit` option of `withInitAction()`
 */

/**
 * The component will render immediately, even if `initAction` is still pending.
 * It is recommended to use this option and render a loading indicator or placeholder content
 * until `initAction` is resolved. This will give the user immediate feedback that something is
 * being loaded. While the `initAction` is pending, an `isInitializing` prop will be passed to
 * the component.
 * @type {string}
 */
export const INIT_SELF_ASYNC = 'ASYNC';

/**
 * This will cause the component not to be mounted until the first initialization has completed.
 * The component will remain mounted during any reinitialization.
 * @type {string}
 */
export const INIT_SELF_BLOCKING = 'BLOCKING';

/**
 * The same behavior as `INIT_SELF_BLOCKING` but it will also unmount the component during
 * reinitialization.
 * @type {string}
 */
export const INIT_SELF_UNMOUNT = 'UNMOUNT';

/**
 * Only execute initAction on the server (initMode == MODE_PREPARE). Initialization will
 * be skipped on the client.
 * @type {string}
 */
export const INIT_SELF_NEVER = 'NEVER';
