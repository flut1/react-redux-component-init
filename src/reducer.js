import { createSelector } from 'reselect';

import createShallowArrayCompareSelector from './utils/createShallowArrayCompareSelector';
import { INIT_COMPONENT, SET_INIT_MODE } from './actions/actionTypes';
import extractValuesForProps from './utils/extractValuesForProps';
import { MODE_PREPARE } from './initMode';

/**
 * Reducer function that manages the state for `react-redux-component-init`. This reducer should
 * be included using Redux's `combineReducers()` under the `init` key in the root of
 * the store state.
 */
export default (state = {
  mode: MODE_PREPARE,
  prepared: {},
  selfInit: {},
}, action) => {
  switch (action.type) {
    case SET_INIT_MODE:
      return {
        ...state,
        mode: action.payload,
      };
    case INIT_COMPONENT:
      if (action.payload.isPrepare) {
        return {
          ...state,
          prepared: {
            ...state.prepared,
            [action.payload.prepareKey]: action.payload.complete,
          },
        };
      }
      return {
        ...state,
        selfInit: {
          ...state.selfInit,
          [action.payload.prepareKey]: action.payload.complete,
        },
      };
    default:
      return state;
  }
};

const createComponentInitValuesSelector = ({
  componentId, initProps, options: { getPrepareKey },
}) =>
  createShallowArrayCompareSelector(
    (state, props) => extractValuesForProps(props, initProps),
    initValues => ({
      prepareKey: getPrepareKey(componentId, initValues),
      initValues,
    }),
  );

export const createComponentInitStateSelector = initConfig => createSelector(
  createComponentInitValuesSelector(initConfig),
  state => state.selfInit,
  state => state.prepared,
  ({ prepareKey, initValues }, selfInit, prepared) => ({
    prepareKey,
    initValues,
    selfInitState: selfInit[prepareKey],
    isPrepared: !!prepared[prepareKey],
  }),
);
