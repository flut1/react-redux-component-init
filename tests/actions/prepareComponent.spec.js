/* global expect, describe, it, jest */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { connect } from 'react-redux';

import withInitAction, { clearComponentIds } from '../../src/withInitAction';
import prepareComponent from '../../src/actions/prepareComponent';
import { MODE_PREPARE } from '../../src/initMode';
import SimpleInitTestComponent from '../fixtures/SimpleInitTestComponent';

const mockStore = configureMockStore([thunk]);

describe('prepareComponent', () => {
  describe('with a component without initConfig', () => {
    it('should return a promise that resolves', () => {
      const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
      return expect(
        store.dispatch(prepareComponent(SimpleInitTestComponent, {})),
      ).resolves.toBeUndefined();
    });
  });
  describe('with a custom getInitState', () => {
    clearComponentIds();
    const store = mockStore({ custom: { mode: MODE_PREPARE, prepared: {} } });
    const MockWithInit = withInitAction(() => Promise.resolve(), {
      getInitState: state => state.custom,
    })(SimpleInitTestComponent);
    const preparePromise = store.dispatch(prepareComponent(MockWithInit, {}));
    it('should dispatch INIT_COMPONENT actions', () =>
      preparePromise.then(() => {
        const actions = store.getActions();
        expect(actions).toMatchSnapshot();
      }));
  });
  describe('with a custom getPrepareKey', () => {
    clearComponentIds();
    const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
    const MockWithInit = withInitAction(() => Promise.resolve(), {
      getPrepareKey: componentId => `${componentId}!foobar`,
    })(SimpleInitTestComponent);
    const preparePromise = store.dispatch(prepareComponent(MockWithInit, {}));
    it('should dispatch INIT_COMPONENT actions with the correct prepareKey', () =>
      preparePromise.then(() => {
        const actions = store.getActions();
        expect(actions).toMatchSnapshot();
      }));
  });
  describe('with a component configured with allowLazy', () => {
    clearComponentIds();
    const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
    const mockInitAction = jest.fn(() => Promise.resolve());
    const MockWithInit = withInitAction(mockInitAction, {
      allowLazy: true,
    })(SimpleInitTestComponent);

    const preparePromise = store.dispatch(prepareComponent(MockWithInit, {}));
    it('should call the initAction', () =>
      preparePromise.then(() => expect(mockInitAction.mock.calls.length).toBe(1)));
    it('should dispatch INIT_COMPONENT actions', () =>
      preparePromise.then(() => {
        const actions = store.getActions();
        expect(actions).toMatchSnapshot();
      }));
  });
  describe('with a component wrapped in connect()', () => {
    clearComponentIds();
    const store = mockStore({ custom: { mode: MODE_PREPARE, prepared: {} } });
    const mockInitAction = jest.fn(() => Promise.resolve());
    const MockWithInit = withInitAction(mockInitAction, {
      getInitState: state => state.custom,
    })(SimpleInitTestComponent);
    const ConnectedMockWithInit = connect(
      null,
      null,
    )(MockWithInit);
    const preparePromise = store.dispatch(prepareComponent(ConnectedMockWithInit, {}));

    it('should dispatch INIT_COMPONENT actions', () =>
      preparePromise.then(() => {
        const actions = store.getActions();
        expect(actions).toMatchSnapshot();
      }));
    it('should call the initAction', () =>
      preparePromise.then(expect(mockInitAction.mock.calls.length).toBe(1)));
  });
  describe('with props', () => {
    clearComponentIds();
    const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
    const mockInitAction = jest.fn(() => Promise.resolve());
    const MockWithInit = withInitAction(['foo', 'bar'], mockInitAction)(SimpleInitTestComponent);
    const preparePromise = store.dispatch(
      prepareComponent(MockWithInit, {
        bar: 'abc',
        foo: 123,
        foobar: 3,
      }),
    );

    it('should only include values of initProps in the prepareKey', () =>
      preparePromise.then(() =>
        expect(store.getActions()[0].payload.prepareKey).toBe('SimpleInitTestComponent[123,"abc"]'),
      ));

    it('should call the initAction with only the initProps values', () =>
      preparePromise.then(() =>
        expect(mockInitAction.mock.calls[0][0]).toEqual({
          bar: 'abc',
          foo: 123,
        }),
      ));
  });
  describe('when not passing one of the values configured in initProps', () => {
    clearComponentIds();
    const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
    const MockWithInit = withInitAction(['foo', 'bar'], () => Promise.resolve())(
      SimpleInitTestComponent,
    );

    it('should throw an error', () =>
      expect(() =>
        store.dispatch(
          prepareComponent(MockWithInit, {
            bar: 'abc',
          }),
        ),
      ).toThrow());
  });
});
