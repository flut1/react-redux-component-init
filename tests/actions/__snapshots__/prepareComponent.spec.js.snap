// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepareComponent with a component configured with allowLazy should dispatch INIT_COMPONENT actions 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
  Object {
    "payload": Object {
      "complete": true,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;

exports[`prepareComponent with a component wrapped in connect() should dispatch INIT_COMPONENT actions 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
  Object {
    "payload": Object {
      "complete": true,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;

exports[`prepareComponent with a custom getInitState should dispatch INIT_COMPONENT actions 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
  Object {
    "payload": Object {
      "complete": true,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;

exports[`prepareComponent with a custom getPrepareKey should dispatch INIT_COMPONENT actions with the correct prepareKey 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent!foobar",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
  Object {
    "payload": Object {
      "complete": true,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent!foobar",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;
